server/asset/eth/rpcclient.go

```diff
@@ -52,111 +56,268 @@ type ethConn struct {
+	// so an ethConn has not been created for them.
+	neverConnectedEndpoints []string
+	healthCheckCounter      int
+	tokensLoaded            map[uint32]bool
+	ethContractAddr         common.Address
+
+	// the order of clients will change based on the health of the connections.
+	clientsMtx sync.RWMutex
+	clients    []*ethConn
 }
 
-func newRPCClient(net dex.Network, endpoints []string, log dex.Logger) *rpcclient {
+func newRPCClient(net dex.Network, endpoints []string, ethContractAddr common.Address, log dex.Logger) *rpcclient {
 	return &rpcclient{
-		net:       net,
-		endpoints: endpoints,
-		log:       log,
+		net:             net,
+		endpoints:       endpoints,
+		log:             log,
+		ethContractAddr: ethContractAddr,
+		tokensLoaded:    make(map[uint32]bool),
+	}
+}
+
+func (c *rpcclient) clientsCopy() []*ethConn {
+	c.clientsMtx.RLock()
+	defer c.clientsMtx.RUnlock()
+
+	clients := make([]*ethConn, len(c.clients))
+	copy(clients, c.clients)
+	return clients
+}
+
+func (c *rpcclient) connectToEndpoint(ctx context.Context, endpoint string) (*ethConn, error) {
+	var success bool
+
+	client, err := rpc.DialContext(ctx, endpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	defer func() {
+		// This shouldn't happen as the only possible errors are due to ETHSwap and
+		// tokener creation.
+		if !success {
+			client.Close()
+		}
+	}()
+
+	ec := &ethConn{
+		Client:   ethclient.NewClient(client),
+		endpoint: endpoint,
+		tokens:   make(map[uint32]*tokener),
+		caller:   client,
+	}
+
+	reqModules := []string{"eth", "txpool"}
+	if err := dexeth.CheckAPIModules(client, endpoint, c.log, reqModules); err != nil {
+		c.log.Warnf("Error checking required modules at %q: %v", endpoint, err)
+		c.log.Warnf("Will not account for pending transactions in balance calculations at %q", endpoint)
+		ec.txPoolSupported = false
+	} else {
+		ec.txPoolSupported = true
+	}
+
+	es, err := swapv0.NewETHSwap(c.ethContractAddr, ec.Client)
+	if err != nil {
+		return nil, fmt.Errorf("unable to initialize eth contract for %q: %v", endpoint, err)
+	}
+	ec.swapContract = &swapSourceV0{es}
+
+	for assetID := range c.tokensLoaded {
+		tkn, err := newTokener(ctx, assetID, c.net, ec.Client)
+		if err != nil {
+			return nil, fmt.Errorf("error constructing ERC20Swap: %w", err)
+		}
+		ec.tokens[assetID] = tkn
+	}
+	success = true
+
+	return ec, nil
+}
+
+type connectionStatus int
+
+const (
+	failed connectionStatus = iota
+	outdated
+	connected
+)
+
+func (c *rpcclient) checkConnectionStatus(ctx context.Context, conn *ethConn) connectionStatus {
+	hdr, err := conn.HeaderByNumber(ctx, nil)
+	if err != nil {
+		c.log.Errorf("Failed to get header from %q: %v", conn.endpoint, err)
+		return failed
+	}
+
+	if c.headerIsOutdated(hdr) {
+		c.log.Warnf("header fetched from %q appears to be outdated (time %s). If you continue to see this message, you might need to check your system clock",
+			conn.endpoint, time.Unix(int64(hdr.Time), 0))
+		return outdated
+	}
```

Comment:
Minor tweak to show age could be nice:

```go
	if c.headerIsOutdated(hdr) {
		hdrTime := time.Unix(int64(hdr.Time), 0)
		c.log.Warnf("header fetched from %q appears to be outdated (time %s is %v old). "+
			"If you continue to see this message, you might need to check your system clock",
			conn.endpoint, hdrTime, time.Since(hdrTime))
		return outdated
	}
```